### PR TITLE
One skinning dispatch per mesh: Part 1

### PR DESCRIPTION
**Pull request by amzn-tommy**
This change splits skinning compute shader dispatches into 1 per mesh instead of 1 per lod. Eventually, it will allow for more flexibility around things like models that have cloth as one sub-mesh (which is skinned on the cpu and thus should be ignored), or has one sub-mesh that is static and not skinned at all, or has a single sub-mesh missing texture coordinates (and thus lacking tangents/bitangents).

This part doesn't modify any functionality, it just splits the dispatch up. It also changes the interface for creating the SkinnedMeshInputBuffers class to accept an Atom model instead of the caller adding a series of individual buffers. The old interface hasn't been removed yet in this PR, but it will be. This change breaks morph targets, which will be updated in a later PR to reduce the size of this already large PR. As such, this is going into a staging branch rather than into development, until cloth and morph targets are brought up to date.

### CODE REVIEW

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshDispatchItem.cpp [hunk 1]
```diff
@@ -27,16 +27,20 @@ namespace AZ
     {
         SkinnedMeshDispatchItem::SkinnedMeshDispatchItem(
```
Comment (amzn-tommy) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshDispatchItem.cpp:28]: This class is being updated to get the input buffers for a specific mesh, rather than for an entire lod.

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshVertexStreamProperties.cpp [hunk 2]
```diff
@@ -27,55 +27,69 @@ namespace AZ
...
                 Name{"m_sourceBlendIndices"},
-                RHI::ShaderSemantic{Name{"UNUSED"}}
+                RHI::ShaderSemantic{Name{"SKIN_JOINTINDICES"}},
+                false, // isOptional
+                SkinnedMeshInputVertexStreams::BlendIndices
             };
 
             m_inputStreamInfo[static_cast<uint8_t>(SkinnedMeshInputVertexStreams::BlendWeights)] = SkinnedMeshVertexStreamInfo{
-                RHI::Format::R32G32B32A32_FLOAT,
+                RHI::Format::R32_FLOAT,
```
Comment (amzn-tommy) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshVertexStreamProperties.cpp:76]: Instead of assuming 4 weights per-vertex, we now have a float buffer that supports any number of influences. This is only one part of the overall pipeline that needs to be updated to support more influences per-vertex.

File: Gems/EMotionFX/Code/EMotionFX/Source/Mesh.cpp [hunk 3]
```diff
@@ -296,25 +295,38 @@ namespace EMotionFX
             skin2dArray.SetNumPreCachedElements(maxSkinInfluences);
             skin2dArray.Resize(modelVertexCount);
 
-            // Fill in skinning data from atom buffer
-            for (uint32 v = 0; v < modelVertexCount; ++v)
+            AZ::u32 currentVertex = 0;            
+            AZ::u32 totalJointIdPadding = 0;
```
Comment (amzn-tommy) [inline Gems/EMotionFX/Code/EMotionFX/Source/Mesh.cpp:299]: Instead of looping over the entire model's vertex count, we now loop over each mesh and increment the overall vertex within the model along the way. This allows us to stop at the end of each mesh and check to see if any padding/dummy jointId's were inserted in the original buffer, in which case we want to skip them while reading data to be added to the EMFX skinningLayer

File: Gems/EMotionFX/Code/EMotionFX/Source/Mesh.cpp [hunk 4]
```diff
@@ -184,11 +185,12 @@ namespace EMotionFX
         // The lod has shared buffers that combine the data from each submesh.
         // These buffers can be accessed through the first submesh in their entirety
         // by using the BufferViewDescriptor from the Buffer instead of the one from the sub-mesh's BufferAssetView
-        const AZ::RPI::ModelLodAsset::Mesh& sourceMesh = sourceModelLod->GetMeshes()[0];
+        const AZ::RPI::ModelLodAsset::Mesh& sourceMesh0 = sourceModelLod->GetMeshes()[0];
```
Comment (amzn-tommy) [inline Gems/EMotionFX/Code/EMotionFX/Source/Mesh.cpp:188]: This was just re-named to prevent a naming conflict further down in the function.

File: Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/ActorAsset.cpp [hunk 5]
```diff
@@ -269,10 +161,10 @@ namespace AZ
                             // The skinned mesh lod gets a unique morph for each meta, since each one has unique min/max delta values to use for decompression
                             const AZStd::string morphString = AZStd::string::format("%s_Lod%zu_Morph_%s", fullFileName.c_str(), lodIndex, metaData.m_meshNodeName.c_str());
 
-                            float minWeight = morphTarget->GetRangeMin();
-                            float maxWeight = morphTarget->GetRangeMax();
+                            [[maybe_unused]]float minWeight = morphTarget->GetRangeMin();
+                            [[maybe_unused]]float maxWeight = morphTarget->GetRangeMax();
 
-                            skinnedMeshLod.AddMorphTarget(metaData, morphBufferAsset, morphString, minWeight, maxWeight);
+                            //skinnedMeshLod.AddMorphTarget(metaData, morphBufferAsset, morphString, minWeight, maxWeight);
```
Comment (amzn-tommy) [inline Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/ActorAsset.cpp:167]: This is intentional to prevent a crash, and will be fixed in the next PR, which is why this change is going into a staging branch instead of development.

File: Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/ActorAsset.cpp [hunk 6]
```diff
@@ -471,49 +314,28 @@ namespace AZ
...
                 // Set colors after morphs are set, so that we know whether or not they are dynamic (if they exist)
-                const RPI::BufferAssetView* colorView = mesh0.GetSemanticBufferAssetView(Name{ "COLOR" });
                 if (colorView)
                 {
                     if (skinnedMeshLod.HasDynamicColors())
                     {
                         // If colors are being morphed,
                         // add them as input to the skinning compute shader, which will apply the morph
-                        skinnedMeshLod.SetSkinningInputBufferAsset(colorView->GetBufferAsset(), SkinnedMeshInputVertexStreams::Color);
+                        //skinnedMeshLod.SetSkinningInputBufferAsset(colorView->GetBufferAsset(), SkinnedMeshInputVertexStreams::Color);
```
Comment (amzn-tommy) [inline Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/ActorAsset.cpp:329]: These are intentional to prevent a crash, and will be fixed in the next PR.

**Review by amzn-tommy — COMMENTED** [review]
Some comments to make the review a bit easier to follow.

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp [hunk 7]
```diff
@@ -57,6 +57,212 @@ namespace AZ
...
+
+            for (const RPI::ModelLodAsset::Mesh::StreamBufferInfo& streamBufferInfo : modelLodAssetMesh.GetStreamBufferInfoList())
+            {
+                // If it is not part of the skinning compute shader input or output, then it is a static buffer used for rendering instead
+                // of skinning
+                bool isStaticStream = !SkinnedMeshVertexStreamPropertyInterface::Get()->GetInputStreamInfo(streamBufferInfo.m_semantic) &&
+                    !SkinnedMeshVertexStreamPropertyInterface::Get()->GetOutputStreamInfo(streamBufferInfo.m_semantic);
+
+                // Also, if it is a color, and colors aren't modified for this mesh, it is static, not dynamic
+                isStaticStream |= streamBufferInfo.m_semantic == RHI::ShaderSemantic{ Name{ "COLOR" } } && m_hasStaticColors;
```
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp:205]: Nit: It's a bit hard to parse a bitwise-or along with other boolean logic happening in one expression at-a-glance.

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp [hunk 8]
```diff
@@ -317,40 +503,41 @@ namespace AZ
             return m_lods[lodIndex].m_inputBuffers[inputStream]->GetBufferView();
         }
 
-        uint32_t SkinnedMeshInputBuffers::GetVertexCount(size_t lodIndex) const
+        bool SkinnedMeshInputBuffers::HasDynamicColors(size_t lodIndex, [[maybe_unused]] size_t meshIndex) const
         {
-            return m_lods[lodIndex].m_vertexCount;
+            // TODO - treat each mesh individually
+            return m_lods[lodIndex].HasDynamicColors();//m_hasDynamicColors[meshIndex];
```
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp:509]: Nit: should this be `HasVertexColors`?

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshVertexStreamProperties.cpp [hunk 9]
```diff
@@ -179,8 +193,38 @@ namespace AZ
...
+                channelInfo.m_isOptional = inputStreamInfo.m_isOptional;
+                channelInfo.m_componentCount = RHI::GetFormatComponentCount(inputStreamInfo.m_elementFormat);
+
+                m_computeShaderInputContract.m_streamChannels.push_back(channelInfo);
+            }
         };
 
+        const SkinnedMeshVertexStreamInfo* SkinnedMeshVertexStreamProperties::GetInputStreamInfo(const RHI::ShaderSemantic& shaderSemantic) const
+        {
+            auto FindVertexStreamInfo = [shaderSemantic](const SkinnedMeshVertexStreamInfo& vertexStreamInfo)
```
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshVertexStreamProperties.cpp:212]: Nit: `&shaderSemantic` avoids a copy

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshVertexStreamProperties.cpp [hunk 10]
```diff
@@ -191,6 +235,24 @@ namespace AZ
             return m_staticStreamInfo[static_cast<uint8_t>(stream)];
         }
 
+        const SkinnedMeshOutputVertexStreamInfo* SkinnedMeshVertexStreamProperties::GetOutputStreamInfo(const RHI::ShaderSemantic& shaderSemantic) const
+        {
+            auto FindVertexStreamInfo = [shaderSemantic](const SkinnedMeshOutputVertexStreamInfo& vertexStreamInfo)
```
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshVertexStreamProperties.cpp:240]: Nit: `&shaderSemantic` avoids a copy

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Model/ModelAssetBuilderComponent.cpp [hunk 11]
```diff
@@ -1355,15 +1356,34 @@ namespace AZ
...
+                    // Pad the joint id buffer if it ends too soon, so the next view can start aligned
+                    uint32_t extraIdCount = CalculateJointIdPaddingCount(newJointIdCount);
 
-                    lodBufferInfo.m_skinInfluencesCount += numNewSkinInfluences;
+                    // Pad the buffer
+                    AZStd::vector<uint16_t> extraIds(extraIdCount, 0);
+                    mesh.m_skinJointIndices.insert(
+                        mesh.m_skinJointIndices.end(), extraIds.begin(), extraIds.end());
+
+                    AZ_Assert(prevJointIdCount * sizeof(uint16_t) % 16 == 0, "Failed to align the joint id offset along a 16-byte boundary");
```
Comment (jeremyong-az) [inline Gems/Atom/RPI/Code/Source/RPI.Builders/Model/ModelAssetBuilderComponent.cpp:1370]: This expression with `* sizeof(uint16_t) % 16 == 0` can't ever evaluate false

File: Gems/Atom/RPI/Code/Source/RPI.Reflect/Model/ModelLodAsset.cpp [hunk 12]
```diff
@@ -26,6 +26,8 @@ namespace AZ
                     ->Version(0)
```
Comment (jeremyong-az) [inline Gems/Atom/RPI/Code/Source/RPI.Reflect/Model/ModelLodAsset.cpp:26]: Do we need a version bump and is any migration needed?

File: Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/ActorAsset.cpp [hunk 13]
```diff
@@ -196,56 +77,67 @@ namespace AZ
...
 
                     // Now that we have the 16-bit indices, pack them into 32-bit uints
+                    uint32_t packedJointIds = 0;
                     for (size_t i = 0; i < localIndices.size(); ++i)
                     {
                         if (i % 2 == 0)
                         {
                             // Put the first/even ids in the most significant bits
-                            blendIndexBufferData[atomVertexBufferOffset + vertexIndex][i / 2] = localIndices[i] << 16;
+                            packedJointIds = localIndices[i] << 16;
```
Comment (jeremyong-az) [inline Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/ActorAsset.cpp:117]: I'm not sure exactly how the compiler would choose to optimize this loop, but I would write it by incrementing the loop counter by 2 each time, and writing `packedJointIds = (localIndices[i] << 16) | localIndices[i];` which I imagine will be a bit more efficient then writing a dirty register, reading it back, then writing it again across loop iterations.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h [hunk 14]
```diff
@@ -197,8 +208,14 @@ namespace AZ
...
-            //! Keep track of the assetId of the asset used to create the SkinnedMeshInputBuffers
-            void SetAssetId(Data::AssetId assetId);
+            //! Create SkinnedMeshInputBuffers from a model
+            void CreateFromModelAsset(const Data::Asset<RPI::ModelAsset>& modelAsset);
+
+            //! Get the ModelAsset used to create the SkinnedMeshInputBuffers
+            Data::Asset<RPI::ModelAsset> GetModelAsset() const;
+
+            //! Get the number of meshes for the lod.
+            size_t GetMeshCount(size_t lodIndex) const;
```
[the remark this answers is not in the document]
Comment (amzn-tommy) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h:218]: Prior to merging back to development, I'm also happy to make the change to signed types across the board in the skinned mesh code where it makes sense if that's the direction we want to take, but since that involves more than the switch from size_t->uint32_t, I'd just as soon save that for a separate pr.

File: Gems/Atom/Feature/Common/Assets/Shaders/SkinnedMesh/LinearSkinningCS.azsl [hunk 15]
```diff
@@ -154,9 +154,9 @@ void MainCS(uint3 thread_id: SV_DispatchThreadID)
     if(i < InstanceSrg::m_numVertices)
     {
         // Moving current vertex position updated last frame to a predefined location to maintain a vertex history between two frames
-        PassSrg::m_skinnedMeshOutputStream[InstanceSrg::m_targetPositions + (InstanceSrg::m_numVertices + i) * 3] = PassSrg::m_skinnedMeshOutputStream[InstanceSrg::m_targetPositions + i * 3];
-        PassSrg::m_skinnedMeshOutputStream[InstanceSrg::m_targetPositions + (InstanceSrg::m_numVertices + i) * 3 + 1] = PassSrg::m_skinnedMeshOutputStream[InstanceSrg::m_targetPositions + i * 3 + 1];
-        PassSrg::m_skinnedMeshOutputStream[InstanceSrg::m_targetPositions + (InstanceSrg::m_numVertices + i) * 3 + 2] = PassSrg::m_skinnedMeshOutputStream[InstanceSrg::m_targetPositions + i * 3 + 2];
+        PassSrg::m_skinnedMeshOutputStream[InstanceSrg::m_targetPositionHistory + i * 3] = PassSrg::m_skinnedMeshOutputStream[InstanceSrg::m_targetPositions + i * 3];
```
[the remark this answers is not in the document]
Comment (jeremyong-az) [reply Gems/Atom/Feature/Common/Assets/Shaders/SkinnedMesh/LinearSkinningCS.azsl:157]: > This is actually one of the reasons I initially started looking into Programmable Vertex Pulling. If there was no binding of various IA buffers for the draw call, and instead each stream was just an offset into a buffer, and we exposed the offsets so that the skinned mesh feature processor can access and modify them each frame, then the skinned mesh feature processor could just swap the offset used by position and position history each frame.

Completely in support of ditching IA... I think it offers little to no value relative to its disadvantages

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (jeremyong-az) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h]: I say we start with the fixed width change and we an revisit the signed type discussion. I have actually spoken to a number of people internally regarding the previous guidance, which I will just say is unfortunate haha

**Review by jeremyong-az — APPROVED** [review]

**Comment by jeremyong-az** [comment]
Sorry about the deleted comment, it was a question I had regarding something you didn't need to change but I was trying to resolve the comment and accidentally deleted it.

**Review by wintermute-motherbrain — APPROVED** [review]
